### PR TITLE
Pouchdb Implementation Revamp

### LINKED ISSUE #2863: Pouch: Restaurants demo fails on step #2 to choose time.
Repro steps:

* Visit web shell with new Pouch user ex. `http://localhost:9119/shells/web-shell/index.html?log=&plannerStorage=volatile&storage=pouchdb://local/foouser`
* Choose `Find restaurants near you.` suggestion and wait for results to load.
* Choose `You are free at 19:00...` suggestion.

Note error as:

```
Uncaught (in promise) Error: can't find store of id !96693423723245:foo--LbyEnYae6T_NtQYj7Rv:21
    at assert (assert-web.ts:12)
    at Arc._instantiateParticle (arc.ts:437)
    at arc.ts:576
    at Array.forEach (<anonymous>)
    at Arc.instantiate (arc.ts:576)
    at Object.onArcLoadRecipe (particle-execution-host.ts:231)
```

There are three earlier errors in the log as:

```
StorageProxy '!96693423723245:foo--LbyEnYae6T_NtQYj7Rv:8' received stale update version 1; current is 1
```

FYI @lindner

### PR DESCRIPTION
**Pull request by lindner**
- Adds a generic Upsert function that abstracts the read-modify-write cycle for Pouch
- Revamps Arc tests to perform most in PouchDB, skip ones that don't
- Add more minor cleanups and tidiness.

Fixes #2863

### CODE REVIEW

**Comment by lindner** [comment]
Making a new PR.

**Comment by lindner** [comment]
Reopening

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  (arg: T): Promise<T>;
+}
+
+/**
+ * Provides a way to read-modify-write data for a given PouchDB key
+ * based on the following rules:
+ *
+ * - A new entry based on defaultValue is stored if it doesn't exist.
+ * - If the existing entry is available it is fetched.
+ * - A copy of the existing document is pass passed to `mutatorFn`
```
Comment (shaper) [inline src/runtime/storage/pouchdb/pouch-db-upsert.ts]: s/pass//

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ * @param defaultValue If the docId does not exist use this and also pass through mutatorFn.
+ */
+export async function upsert<T extends UpsertDoc>(
+    db: PouchDB.Database,
+    docId: string,
+    mutatorFn?: UpsertMutatorFn<T>,
+    defaultValue?: PouchDB.Core.NewDocument<T>): Promise<T> {
+
+  // Keep retrying the operation until it succeeds.
+  // TODO(lindner): add backoff and error out if this goes on for too long
```
Comment (shaper) [inline src/runtime/storage/pouchdb/pouch-db-upsert.ts]: Perhaps at least add a count and bail within 10 - 100 tries to at least avoid infinite loop risk until doing backoff.

File: src/runtime/test/artifacts/test-particle.js [hunk 3]
```diff
@@ -14,6 +14,7 @@ defineParticle(({Particle}) => {
       this.barHandle = handles.get('bar');
     }
     onHandleSync(handle, model) {
+      console.log('>onHandleSync');
```
Comment (shaper) [inline src/runtime/test/artifacts/test-particle.js]: rm?

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {PouchDB, PouchDbDebug, PouchDbMemory} from '../../../platform/pouchdb-web.js';
+
+export interface UpsertDoc {
+  version: number; // TODO version
```
Comment (shaper) [inline src/runtime/storage/pouchdb/pouch-db-upsert.ts]: What does `TODO version` mean here?

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      doc = await db.get(docId);
+      existingDoc = true;
+    } catch (err) {
+      if (err.name !== 'not_found') {
+        throw err;
+      }
+      doc = {_rev: undefined, _id: docId, ...defaultValue};
+    }
+
+    // TODO deepClone? Readonly?  For now just use a shallow clone to avoid weirdness..
```
Comment (shaper) [inline src/runtime/storage/pouchdb/pouch-db-upsert.ts]: What weirdness do you have in mind? To help next reader.

File: src/runtime/storage/pouchdb/pouch-db-storage.ts [hunk 6]
```diff
@@ -96,6 +100,7 @@ export class PouchDbStorage extends StorageBase {
       return this.construct(id, type, key);
     } catch (err) {
       if (err.name && err.name === 'not_found') {
+        console.warn('connecting despite missing doc, returning null');
```
Comment (shaper) [inline src/runtime/storage/pouchdb/pouch-db-storage.ts]: Is it reasonable to return null here? Maybe throwing an error would make more sense?

It looks like Firebase equivalent should never return null, and throws errors in earlier sanity-checking of input params before ending up in `this.getInstance(key).storage;` and `storage` looks like it is expected to never be null.

**Comment by lindner** [comment]
This is ready to be reviewed.

Probably best to start with `pouch-db-upsert.ts` which factors out the read-modify-write that was implemented in 4 places.  It should have it's own tests independent of the general  `pouchdb-tests.ts`

There are some tricky areas around referenceMode handling, and there's a very odd failure in `arc-tests.ts` when using a volatile storage driver that I'm looking into.

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts (hunk 5 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-upsert.ts]: I'll document this more.   The mutator receives an object that is modifiable; but after the call there's a check to determine if changes were made.  If the mutator modifies the original object then that will fail.

So we pass in a clone of the object.  Using the spread operator only does a shallow clone though.  In future a deepClone would be preferred to allow the mutator to modify the object any way it seems fit.

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts (hunk 4 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-upsert.ts]: obsolete comment, will remove.

File: src/runtime/test/artifacts/test-particle.js (hunk 3 above)
Comment (lindner) [reply src/runtime/test/artifacts/test-particle.js]: ack

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts (hunk 2 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-upsert.ts]: Ack, working on it.

File: src/runtime/storage/pouchdb/pouch-db-storage.ts (hunk 6 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-storage.ts]: I've just restructured this so I think we'll be able to do this.

Right now the only place I see this happen is shell requesting the launcher arc.

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts (hunk 2 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-upsert.ts]: done

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts (hunk 1 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-upsert.ts]: done

File: src/runtime/storage/pouchdb/pouch-db-storage.ts (hunk 6 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-storage.ts]: comment removed, only used in connectOrConstruct

File: src/runtime/storage/pouchdb/pouch-db-upsert.ts (hunk 5 above)
Comment (lindner) [reply src/runtime/storage/pouchdb/pouch-db-upsert.ts]: docs updated.

**Comment by lindner** [comment]
There was one missing await this.initialized in fromLiteral that was causing instability and caused referenceMode to be reset.  Now fixed.

Manual and automated tests passing, going to submit this.

**Pull request merged**